### PR TITLE
feat(controller): Enhanced TTL controller scalability

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: docs/scaling.md [hunk 1]
```diff
@@ -10,8 +10,7 @@ You cannot horizontally scale the controller.
 
 You can scale the controller vertically:
 
-- If you have workflows with many steps, increase `--pod-workers`.
```
Comment (alexec) [inline docs/scaling.md:13]: I don't recommend changing `--pod-workers` anymore. It has no impact in any testing I've done.

File: workflow/ttlcontroller/ttlcontroller.go [hunk 2]
```diff
@@ -129,9 +130,6 @@ func (c *Controller) enqueueWF(obj interface{}) {
 		log.Warnf("'%v' is not an unstructured", obj)
 		return
 	}
-	if un.GetDeletionTimestamp() != nil {
```
Comment (alexec) [inline workflow/ttlcontroller/ttlcontroller.go:132]: redundant now it is checked by `FilterFunc` on line 52

File: workflow/ttlcontroller/ttlcontroller.go [hunk 3]
```diff
@@ -187,9 +192,14 @@ func (c *Controller) deleteWorkflow(key string) error {
 
 		err = c.wfclientset.ArgoprojV1alpha1().Workflows(wf.Namespace).Delete(wf.Name, &metav1.DeleteOptions{PropagationPolicy: commonutil.GetDeletePropagation()})
 		if err != nil {
-			return err
+			if apierr.IsNotFound(err) {
```
Comment (alexec) [inline workflow/ttlcontroller/ttlcontroller.go:195]: I've not seen this error apart from initial testing - prevented by line 155 I think, but a good bonus check

File: workflow/ttlcontroller/ttlcontroller.go [hunk 4]
```diff
@@ -49,28 +51,29 @@ func NewController(wfClientset wfclientset.Interface, wfInformer cache.SharedInd
...
 			un, ok := obj.(*unstructured.Unstructured)
-			return ok && un.GetLabels()[common.LabelKeyCompleted] == "true" && un.GetLabels()[common.LabelKeyWorkflowArchivingStatus] != "Pending"
+			return ok && un.GetDeletionTimestamp() == nil && un.GetLabels()[common.LabelKeyCompleted] == "true" && un.GetLabels()[common.LabelKeyWorkflowArchivingStatus] != "Pending"
 		},
 		Handler: cache.ResourceEventHandlerFuncs{
 			AddFunc: controller.enqueueWF,
 			UpdateFunc: func(old, new interface{}) {
 				controller.enqueueWF(new)
 			},
-			DeleteFunc: controller.enqueueWF,
```
Comment (alexec) [inline workflow/ttlcontroller/ttlcontroller.go:59]: this is invoked after the WF is deleted - so it is not needed - removing reduces the number of API requests by 1

File: workflow/ttlcontroller/ttlcontroller_test.go [hunk 5]
```diff
@@ -363,23 +363,24 @@ func TestEnqueueWF(t *testing.T) {
...
 	assert.Equal(t, 0, controller.workqueue.Len())
 
 	// Verify we enqueue when ttl is expired
 	wf.Spec.TTLSecondsAfterFinished = &ten
 	wf.Status.FinishedAt = metav1.Time{Time: controller.clock.Now().Add(-11 * time.Second)}
 	un, err = util.ToUnstructured(wf)
 	assert.NoError(t, err)
-	controller.enqueueWF(un)
+	enqueueWF(controller, un)
+	time.Sleep(2 * time.Second)
```
Comment (alexec) [inline workflow/ttlcontroller/ttlcontroller_test.go:383]: delete this line

File: test/e2e/functional_test.go [hunk 6]
```diff
@@ -88,6 +88,30 @@ func (s *FunctionalSuite) TestDeletingRunningPodWithOrErrorRetryPolicy() {
...
+  entrypoint: main
+  templates:
+    - name: main
+      container:
+        image: argoproj/argosay:v2
+`).
+		When().
+		SubmitWorkflow().
+		WaitForWorkflow().
+		Wait(3 * time.Second). // enought time for TTl controller to delete the workflow
```
Comment (alexec) [inline test/e2e/functional_test.go:110]: correct the capitalization:  "TTl" to "TTL"

File: workflow/ttlcontroller/ttlcontroller.go [hunk 7]
```diff
@@ -152,7 +152,11 @@ func (c *Controller) enqueueWF(obj interface{}) {
 		runtimeutil.HandleError(err)
 		return
 	}
-	//c.workqueue.Add(key)
+	// if we try and delete in the next second, it is possible that the informer is out of sync, our double-check that
```
Comment (alexec) [inline workflow/ttlcontroller/ttlcontroller.go:155]: change "possible" to "almost certain"

File: workflow/ttlcontroller/ttlcontroller_test.go [hunk 8]
```diff
@@ -396,15 +397,17 @@ func TestTTLStrategySucceded(t *testing.T) {
 	wf.Status.FinishedAt = metav1.Time{Time: controller.clock.Now().Add(-5 * time.Second)}
 	un, err = util.ToUnstructured(wf)
 	assert.NoError(t, err)
-	controller.enqueueWF(un)
+	time.Sleep(2 * time.Second)
```
Comment (alexec) [inline workflow/ttlcontroller/ttlcontroller_test.go:400]: delete this extra line

**Review by sarabala1979 — APPROVED** [review]

**Pull request merged**